src/gt4py/stencil_builder.py

```diff
@@ -0,0 +1,210 @@
+
+        *args: Any
+            Passed through to the caching strategy factory
+
+        **kwargs: Any
+            Passed through to the caching strategy factory
+
+        Returns
+        -------
+        type: :py:class:`StencilBuilder`
+            Return self
+
+        Resets all cached build data.
+        """
+        self._build_data = {}
+        self.caching = gt4py.caching.strategy_factory(caching_strategy_name, self, *args, **kwargs)
+        return self
+
+    @property
+    def definition(self) -> AnnotatedStencilFunc:
+        """Get the annotated stencil definition."""
+        return self._build_data.get("prepared_def", self._make_prepared_def())
+
+    def _make_prepared_def(self) -> Callable:
+        """Annotate the stencil definition."""
+        self._build_data["prepared_def"] = self.frontend.prepare_stencil_definition(
+            self._definition, self.externals
+        )
+        assert self._build_data["prepared_def"]._gtscript_
+        return self._build_data["prepared_def"]
+
+    @property
+    def externals(self) -> Dict[str, Any]:
+        """The externals for this build."""
+        return self._build_data.get("externals", {})
+
+    def with_externals(self: SelfType, externals: Dict[str, Any]) -> SelfType:
+        """
+        Fluidly set externals for this build.
+
+        Resets all cached build data.
+        """
+        self._build_data = {}
+        self._build_data["externals"] = externals
+        return self
+
+    @property
+    def backend_data(self) -> Dict[str, Any]:
+        """Get build data set by the backend."""
+        return self._build_data.get("backend_data", {}).copy()
+
+    def with_backend_data(self: SelfType, data: Dict[str, Any]) -> SelfType:
+        """Set backend specific build data."""
+        self._build_data["backend_data"] = {**self.backend_data, **data}
+        return self
+
+    @property
+    def stencil_id(self) -> StencilID:
+        """Get (cached) stencil id."""
+        return self._build_data.get("id", self._make_stencil_id())
+
+    def _make_stencil_id(self) -> StencilID:
+        """Get the stencil id from the caching strategy and cache it."""
+        self._build_data["id"] = self.caching.stencil_id
+        return self._build_data["id"]
+
+    @property
+    def root_pkg_name(self) -> str:
+        """Get the (cached) root package name."""
+        return self._build_data.get("root_pkg_name", self._make_root_pkg_name())
+
+    def _make_root_pkg_name(self) -> str:
+        """Get the root package name from settings and cache it."""
+        self._build_data["root_pkg_name"] = gt4py.config.code_settings["root_package_name"]
+        return self._build_data["root_pkg_name"]
+
+    def with_root_pkg_name(self: SelfType, name: str) -> SelfType:
+        self._build_data["root_pkg_name"] = name
+        return self
+
+    @property
+    def pkg_name(self) -> str:
+        """Get the package name for this stencil."""
```

Comment:
I have a strong opinion about such comments. I don't know what the python world thinks...

Reply:
I agree in general (and I don't think this depends a lot on the programming language).

However, I see a problem when setting the linters for the docstrings: we either set the linters to not to be overly picky with documenting all the the public methods, or we set a strict configuration and then use comments to disable the checks in all trivial cases (which might be a lot). I'm not really sure which one is worse (and don't like any of them in particular).